api/src/public/login/script.js

```diff
@@ -177,14 +177,39 @@ const getUserCtx = function() {
   }
 };
 
-const checkSession = function() {
+const requestSession = () => {
+  return new Promise((resolve, reject) => {
+    request('GET', '/_session', null, xmlhttp => {
+      if (xmlhttp.status === 200) {
+        let response;
+        try {
+          response = JSON.parse(xmlhttp.responseText);
+        } catch (error) {
+          // No body
+        }
+        return resolve(response);
+      }
+      return reject(xmlhttp);
+    });
+  });
+};
+
+const checkSession = async () => {
   if (getCookie('login') === 'force') {
-    // require user to login regardless of session state
+    // Require user to login regardless of session state
     return;
   }
-  const userCtx = getUserCtx();
+
+  let userCtx;
+  if (window.navigator.onLine) {
+    const session = await requestSession();
+    userCtx = session && session.userCtx;
+  } else {
+    userCtx = getUserCtx();
+  }
```

Comment:
Unfortunately `window.navigator.onLine` is not reliable enough for an offline-first app, in part because all browsers implement it differently, but also because an incredibly slow connection will report as `onLine` but the request effectively never complete.

Because the calling function doesn't `await` this means if the user has a userCtx cookie and has a very slow connection, they will still be presented with the login page and have to log in manually.

Reply:
Thank you!
I think the safest way is to always make the request to get the server's session cookie, if it returns a status code different than 404 or 401 then check for the browser `userCtx`.

Reply:
Right... but you can't make a request first in an offline first app.

Reply:
Yeah, however JavaScript doesn't gives us more options to reliable tell if the user is online/offline. 

We do use the `window.navigator.online` for [syncing the DB](https://github.com/medic/cht-core/blob/ebd683d3a8efcef93bae77e6f37834118a855402/webapp/src/ts/services/db-sync.service.ts#L79) which will keep us in the offline first principle, but not always reliable. 

If I were to pick one, I'd pick reliability (request first), but if we think the principle is more important then I think we need to keep the `window.navigator.online`

Reply:
I did a quick research but can't find a solution that won't make a request first underneath

Reply:
Making the request first isn't reliable because the request may not complete (or may take minutes) if the user has a slow connection. That's why the offline first principle is important - it makes the experience reliable for the user.

This is why this code was designed to trust the userCtx and load the app assuming the user is online. If there is an infinite loop happening that definitely needs to be fixed, but we need to find a different way.

If you like, book me in for a call tomorrow sometime and we can go over alternatives.

Reply:
Today, I've been trying to find where the redirect back to login is happening, I don't think it's in the Android apps (prob not loading before redirecting), so I've been checking the api but no luck yet. I'll continue tomorrow.

Reply:
I found the culprit. The api's [server-utils.js is redirecting](https://github.com/medic/cht-core/blob/23cf129a7c0674ec57f2a3f8d340d176c56e46fb/api/src/server-utils.js#L96) back to login, I've added the `res.cookie('login', 'force', {});` before it redirects in this line to solve the problem.
I'm going to polish and add some unit test before review.

Reply:
Done